File: 백준/Gold/9251. LCS/LCS.py
import sys
input = sys.stdin.readline

str1 = input().rstrip()
str2 = input().rstrip()
n1 = len(str1)
n2 = len(str2)

#주의. 문제에서 두 문자열의 길이가 같다고 한 적 없다.
dp = [[0] * (n2+1) for _ in range(n1+1)]  #2차원 리스트로, 두 문자열의 문자가 각 각 행과 열을 가리킴킴
for i in range(1, n1+1):
    for j in range(1, n2+1):
        if str1[i-1] == str2[j-1]:  #같은 문자인 경우, LCS가 하나 증가함함
            dp[i][j] = dp[i-1][j-1] + 1
        else:  #다른 문자인 경우, 새로 추가된 문자에 대해 각 각의 LCS 중 더 큰 것으로 가져감.
            dp[i][j] = max(dp[i-1][j], dp[i][j-1])

print(dp[n1][n2])

File: 백준/Gold/9251. LCS/LCS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

//DP - LCS 복습
public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String str1 = br.readLine();
        String str2 = br.readLine();

        int[][] dp = new int[str1.length() + 1][str2.length() + 1];
        for (int i = 1; i < str1.length() + 1; i++) {
            for (int j = 1; j < str2.length() + 1; j++) {
                if (str1.charAt(i - 1) == str2.charAt(j - 1)) {  //i번째 문자와 j번째 문자가 같은 경우 -> 증가
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                } else {  //다를 경우, 더 큰 값으로
                    dp[i][j] = Math.max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
        System.out.println(dp[str1.length()][str2.length()]);
    }
}
